### ISSUE TITLE
`DoesNotReturnIfAttribute` does not work at all

### ISSUE THREAD

**Issue by Uzivatel919** [dotnet/roslyn#55105]
**Reference**

- [DoesNotReturnIfAttribute Class](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.doesnotreturnifattribute?view=net-5.0)
- [Verify unreachable code](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#verify-unreachable-code)

**Version Used**: 

[Visual Studio Info.txt](https://github.com/dotnet/roslyn/files/6872894/Visual.Studio.Info.txt)

**Sample**:
```c#
class UnreachableCodeTests
{
  object _field = new ();
  bool _isInitialized = false;

  private static void FailFastIf ( [DoesNotReturnIf (false)] bool isValid )
  {
    if ( !isValid )
    {
      throw new InvalidOperationException ();
    }
  }

  public void SetFieldState ( object? containedField )
  {
    _isInitialized = false;

    FailFastIf (_isInitialized);

    _field = containedField; // Warning CS8601 Possible null reference assignment.      
  }
}
```

**Expected Behavior**:
`DoesNotReturnIfAttribute` works as expected.

**Actual Behavior**:
`DoesNotReturnIfAttribute` does not work at all. It must be detected that `_field` assignation is not possible (in regard to annotation stated by `DoesNotReturnIf(false)`) thus `Warning CS8601` in `#nullable enable` context must not be emitted.

### Sample project
[DoesNotReturnTests.zip](https://github.com/dotnet/roslyn/files/6872911/DoesNotReturnTests.zip)

**Comment by RikkiGibson** [comment]
The issue here is that the analysis doesn't see `_isInitialized` as a constant `false` when it is passed to `FailFastIf`. Therefore we don't see the state after the `FailFastIf()` call as a unreachable state.

We could try to augment flow analysis to track whether a non-`const` variable contains a constant value, for example, but I don't expect us to invest in this at least in the short term.

I would consider the observed behavior to be "by design".

**Comment by Uzivatel919** [comment]
I verified that for `const` `bool` it works. Slightly unnaturally for `readonly` `bool` same does not apply. I see it is just ways-and-means issue. Not any failure.
